### PR TITLE
feat: fluentbit hotreload

### LINKED ISSUE #1312: Fluentbit config hot reload
**Is your feature request related to a problem? Please describe.**
Fluentbit restarts can take a long time if there is lot of files to index. 

**Describe the solution you'd like**
Integration of the fluentbit new hot reload feature, see
https://fluentbit.io/announcements/v2.1.0/

### PR DESCRIPTION
**Pull request by aslafy-z**
Enable [fluentbit hotreload feature](https://docs.fluentbit.io/manual/administration/hot-reload) behind a feature flag (FluentbitSpec.HotReload).

I'd say it misses an E2E test. Please do an initial review first. :pray:  

Closes #1312

### CODE REVIEW

**Comment by pepov** [comment]
Looks good, but I would be conservative to keep at least the option of switching back to the legacy method.

File: pkg/resources/fluentbit/daemonset.go [hunk 1]
```diff
@@ -147,6 +154,44 @@ func (r *Reconciler) generatePortsMetrics() (containerPorts []corev1.ContainerPo
 	return
 }
 
+func newConfigMapReloader(spec *v1beta1.FluentbitSpec) *corev1.Container {
```
Comment (pepov) [inline pkg/resources/fluentbit/daemonset.go:157]: can't we simply return by value?

File: pkg/sdk/logging/api/v1beta1/fluentbit_types.go [hunk 2]
```diff
@@ -80,13 +81,14 @@ type FluentbitSpec struct {
...
-	CoroStackSize int32                       `json:"coroStackSize,omitempty" plugin:"default:24576"`
-	Resources     corev1.ResourceRequirements `json:"resources,omitempty"`
-	Tolerations   []corev1.Toleration         `json:"tolerations,omitempty"`
-	NodeSelector  map[string]string           `json:"nodeSelector,omitempty"`
-	Affinity      *corev1.Affinity            `json:"affinity,omitempty"`
-	Metrics       *Metrics                    `json:"metrics,omitempty"`
-	Security      *Security                   `json:"security,omitempty"`
+	CoroStackSize           int32                       `json:"coroStackSize,omitempty" plugin:"default:24576"`
+	Resources               corev1.ResourceRequirements `json:"resources,omitempty"`
+	ConfigReloaderResources corev1.ResourceRequirements `json:"configReloaderResources,omitempty"`
```
Comment (pepov) [inline pkg/sdk/logging/api/v1beta1/fluentbit_types.go:86]: I would consider a different structure here:
```
configHotReload: # pointer, which means hotreload is enabled if set
  image:
  resources:
```

Comment (aslafy-z) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go:86]: I replicated the pattern used for aggregators. Do you want to migrate them to this spec too?

Comment (pepov) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go:86]: I would not touch that, but I also don't want to add 3 new fields to the already big fluentbit spec. I understand that consistency is a nice thing, but I would rely on the user having to read the docs in this case.

File: pkg/resources/fluentbit/daemonset.go (hunk 1 above)
Comment (aslafy-z) [reply pkg/resources/fluentbit/daemonset.go]: Done

File: pkg/sdk/logging/api/v1beta1/fluentbit_types.go (hunk 2 above)
Comment (aslafy-z) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: How can I prevent kubebuilder from adding these subfields into the CRD?

Comment (pepov) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: which one specifically?

Comment (aslafy-z) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: Content of the HotReload struct

File: pkg/resources/fluentbit/daemonset.go [hunk 3]
```diff
@@ -147,6 +154,44 @@ func (r *Reconciler) generatePortsMetrics() (containerPorts []corev1.ContainerPo
...
+	var args []string
+	vm := []corev1.VolumeMount{
+		{
+			Name:      "config",
+			MountPath: OperatorConfigPath,
+		},
+	}
+
+	args = append(args,
+		"--volume-dir=/fluentbit/config",
```
Comment (pepov) [inline pkg/resources/fluentbit/daemonset.go:167]: shouldn't this be `OperatorConfigPath`?

File: pkg/sdk/logging/api/v1beta1/fluentbit_types.go (hunk 2 above)
Comment (pepov) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: I don't think you can reference an object in a crd spec, do you have an example where you have seen that working?

Comment (aslafy-z) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: > I also don't want to add 3 new fields to the already big fluentbit spec
Any idea how I can skip these fields in the crd?

Comment (pepov) [reply pkg/sdk/logging/api/v1beta1/fluentbit_types.go]: I meant I would avoid adding 3 fields on the top level. This is all good as it is now.

File: pkg/resources/fluentbit/daemonset.go (hunk 3 above)
Comment (pepov) [reply pkg/resources/fluentbit/daemonset.go:167]: @aslafy-z are you sure this is the right path?

Comment (aslafy-z) [inline pkg/resources/fluentbit/daemonset.go:167]:
```suggestion
		fmt.Sprintf("--volume-dir=%s", OperatorConfigPath),
```

Comment (aslafy-z) [reply pkg/resources/fluentbit/daemonset.go]: Nice catch. It's still untested!

**Comment by pepov** [comment]
Since LoggingRoute is now in, we could test this by adding a LoggingRoute for a specific Logging resource as we go and check it being applied live.

**Comment by aslafy-z** [comment]
@pepov I added an E2E test. Can you please give a look? :)

File: pkg/resources/fluentbit/daemonset.go [hunk 4]
```diff
@@ -58,6 +58,9 @@ func (r *Reconciler) daemonSet() (runtime.Object, reconciler.DesiredState, error
 	containers := []corev1.Container{
 		*r.fluentbitContainer(),
 	}
+	if r.fluentbitSpec.ConfigHotReload != nil {
+		containers = append(containers, newConfigMapReloader(r.Logging.Spec.FluentbitSpec))
```
Comment (aslafy-z) [inline pkg/resources/fluentbit/daemonset.go:62]:
```suggestion
		containers = append(containers, newConfigMapReloader(r.fluentbitSpec))
```

File: e2e/fluentbit-hotreload/fluentbit_hotreload_test.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+		},
+	}
+	common.RequireNoError(t, c.Create(ctx, &flow))
+
+	agent := v1beta1.FluentbitAgent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "infra",
+		},
+		Spec: v1beta1.FluentbitSpec{
+			LoggingRef: "infra",
```
Comment (aslafy-z) [inline e2e/fluentbit-hotreload/fluentbit_hotreload_test.go:212]:
```suggestion
			LoggingRef: "infra",
			ConfigHotReload: map[string]string{},
```

File: e2e/fluentbit-hotreload/fluentbit_hotreload_test.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
+// Copyright © 2021 Cisco Systems, Inc. and/or its affiliates
```
Comment (pepov) [inline e2e/fluentbit-hotreload/fluentbit_hotreload_test.go:1]:
```suggestion
// Copyright © 2024 Kube logging authors, Inc. and/or its affiliates
```

**Review by pepov — COMMENTED** [review]
Hugo kudos for the test! I would like to add a few things there if you don't mind:
- check that the fluentbit daemonset has been start
- check that no logs are coming before adding the loggingroute
- check if the daemonset has the same generation after adding the loggingroute

I'm happy to put these on top of your branch, just let me know if you have any concerns.

**Review by pepov — APPROVED** [review]

**Review by OverOrion — APPROVED** [review]
LGTM, thanks for this! :tada: :raised_hands:

**Pull request merged**